File: CodingInterviewPatterns/3.Linked_Lists/1.Linked_List_Reversal/Iterative/linkedListReversal.js
/*

Reverse a singly linked list.

1 -> 2 -3 -> 4

     to

4 -> 3 -> 2 -> 1

*/


import { ListNode } from './ds.js'
/**
 * Definition of ListNode:
 * class ListNode {
 *     constructor(val = null, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

export function linked_list_reversal(head) {

  // Declare two pointers at head and null
  let prev = null, curr = head;

  // While we have not reached the end of the list continue
  while (curr != null) {
    // Store the next node before break the link
    let nextNode = curr.next;
    // Reverse the link of the node
    curr.next = prev

    // Move forward prev and curr
    prev = curr;
    curr = nextNode;
  }

  // Return the new head which is the last node
  return prev;
}

File: CodingInterviewPatterns/3.Linked_Lists/1.Linked_List_Reversal/Iterative/LinkedListReversal.java
/*

Reverse a singly linked list.

1 -> 2 -> 3 -> 4

      to

4 -> 3 -> 2 -> 1

*/


import core.LinkedList.ListNode;
/**
 * Definition of ListNode:
 * public class ListNode<T> {
 *     T val;
 *     ListNode next;
 *     ListNode(T val) {
 *         this.val = val;
 *         this.next = null;
 *     }
 * }
 */
 
public class Main {
    public static ListNode<Integer> linked_list_reversal(ListNode<Integer> head) {

        // Declare to pointers at null and head
        ListNode<Integer> prev = null, curr = head;

        // While we haven´t reached the last node
        while (curr != null) {
          
            // Save the next node before break the link
            ListNode<Integer> nextNode = curr.next;
            // Reverse the pointer of current node
            curr.next = prev;

            // Move forward prev to curr and curr to next node
            prev = curr;
            curr = nextNode;
        }

        // Return the new head which is last node
        return prev;
    }
}


// Time complexity: O(n)
